src/libstd/sys/sgx/rwlock.rs

```diff
@@ -1,288 +0,0 @@
-        let rguard = try_lock_or_false!(self.readers);
-        let mut wguard = try_lock_or_false!(self.writer);
-        if *wguard.lock_var() || rguard.lock_var().is_some() {
-            // Another thread has the lock
-            false
-        } else {
-            // We are just now obtaining the lock
-            *wguard.lock_var_mut() = true;
-            true
-        }
-    }
-
-    #[inline]
-    unsafe fn __read_unlock(
-        &self,
-        mut rguard: SpinMutexGuard<'_, WaitVariable<Option<NonZeroUsize>>>,
-        wguard: SpinMutexGuard<'_, WaitVariable<bool>>,
-    ) {
-        *rguard.lock_var_mut() = NonZeroUsize::new(rguard.lock_var().unwrap().get() - 1);
-        if rguard.lock_var().is_some() {
-            // There are other active readers
-        } else {
-            if let Ok(mut wguard) = WaitQueue::notify_one(wguard) {
-                // A writer was waiting, pass the lock
-                *wguard.lock_var_mut() = true;
-            } else {
-                // No writers were waiting, the lock is released
-                assert!(rguard.queue_empty());
-            }
-        }
-    }
-
-    #[inline]
-    pub unsafe fn read_unlock(&self) {
-        let rguard = self.readers.lock();
-        let wguard = self.writer.lock();
-        self.__read_unlock(rguard, wguard);
-    }
-
-    #[inline]
-    unsafe fn __write_unlock(
-        &self,
-        rguard: SpinMutexGuard<'_, WaitVariable<Option<NonZeroUsize>>>,
-        wguard: SpinMutexGuard<'_, WaitVariable<bool>>,
-    ) {
-        if let Err(mut wguard) = WaitQueue::notify_one(wguard) {
-            // No writers waiting, release the write lock
-            *wguard.lock_var_mut() = false;
-            if let Ok(mut rguard) = WaitQueue::notify_all(rguard) {
-                // One or more readers were waiting, pass the lock to them
-                if let NotifiedTcs::All { count } = rguard.notified_tcs() {
-                    *rguard.lock_var_mut() = Some(count)
-                } else {
-                    unreachable!() // called notify_all
-                }
-            } else {
-                // No readers waiting, the lock is released
-            }
-        } else {
-            // There was a thread waiting for write, just pass the lock
-        }
-    }
-
-    #[inline]
-    pub unsafe fn write_unlock(&self) {
-        let rguard = self.readers.lock();
-        let wguard = self.writer.lock();
-        self.__write_unlock(rguard, wguard);
-    }
-
-    // only used by __rust_rwlock_unlock below
-    #[inline]
-    #[cfg_attr(test, allow(dead_code))]
-    unsafe fn unlock(&self) {
-        let rguard = self.readers.lock();
-        let wguard = self.writer.lock();
-        if *wguard.lock_var() == true {
-            self.__write_unlock(rguard, wguard);
-        } else {
-            self.__read_unlock(rguard, wguard);
-        }
-    }
-
-    #[inline]
-    pub unsafe fn destroy(&self) {}
-}
-
-#[cfg(not(test))]
```

Comment:
I think I need some more context to know how, where and why they need to be kept. Let's break it down into categories:

* The functions taking `*mut RWLock` arguments, what would they take instead now when that type does not exist. And what are those functions supposed to be used for?

* Some functions does not appear to be rwlock related: `__rust_print_err`, `__rust_abort`, `__rust_c_alloc` and `__rust_c_dealloc`. Is that so? Why were they in this file to begin with and does it matter where I re-introduce them?

* I suppose you mean down to `mod tests {` and no further? Since the tests are for a type that no longer exists.

Reply:
> The functions taking `*mut RWLock` arguments, what would they take instead now when that type does not exist. And what are those functions supposed to be used for?

These functions are used by libunwind (as mentioned in the code). libunwind is written in C and expects a pthread-style interface to an rwlock. I'm not familiar with the `parking_lot` internals but I suppose this should be a `RawRwLock`? Note that you need to implement an `unlock` function that can unlock from either a reader or a writer state, this is not something that is required by normal Rust.

> Some functions does not appear to be rwlock related: `__rust_print_err`, `__rust_abort`, `__rust_c_alloc` and `__rust_c_dealloc`. Is that so? Why were they in this file to begin with and does it matter where I re-introduce them?

Indeed they're not related. There's a cyclic dependency between `libstd.rlib` and `libunwind.a`. The Rust linking model doesn't support this well. `libunwind` is linked after `libstd`, so the linker must've already decided to include the CGU with these symbols when it's processing `libunwind` for the link to succeed. They're in this file because this increases the likelyhood that the CGU with these symbols will be included. The CGU that includes RWLock is likely to be included in the link.

> I suppose you mean down to `mod tests {` and no further? Since the tests are for a type that no longer exists.

No, as you're changing the RWLock type, the test (and the corresponding C header that's being tested) must be updated to match.

Reply:
I think I understand the problem space a bit more now. Your libunwind port needs access to *a* rwlock implementation to function correctly. It does not have to be any specific implementation, just one that works. Then I get the relationship of these functions and the previous `RWLock` type. Am I interpreting this correctly?

Then we could just have some `sgx_libunwind_stuff.rs` (not serious name suggestion) anywhere (as long as it's included correctly for the linking) and put everything that your libunwind needs there. If I understand it correctly these exposed functions are not really `RWLock` related per se, they are just all the dependencies your libunwind port needs:
* A read-write lock it can use
* A way to print to stderr
* A way to abort
* And a way to alloc/dealloc memory

Reply:
This sounds about right.

Reply:
There is no `__rust_rwlock_new` function or similar exposed. But I checked your fork of `libunwind` now. I assume we have to adjust the `RWLOCK_INIT` constant in that library code to be equal to what `parking_lot::RawRwLock::INIT` is? Which happens to be as simple as an `usize` that is zero.

Reply:
Right.

Reply:
Could you please check my testing branch if you think this looks reasonable, before I push it into this PR: https://github.com/faern/rust/blob/add-parking-lot-test/src/libstd/sys/sgx/mod.rs#L153

Reply:
Why do you need the additional `state` field? Doesn't the `RawRwLock` record what state it's in? You may assume 64-bit.

I just had an idea that would simplify the linking issue I was talking about. If I can get it to work you can avoid the non-rwlock related impls in your code. Will cc you if I can come up with a PR that works.

Reply:
Also I think you should add `[cfg(feature = "i-am-libstd"))]`

Reply:
`RawRwLock` does indeed keep the state internally. But it does not expose the state. Neither does it expose any `unlock` method that safely unlocks either the read or write, depending on which state it's currently in. So I added this extra type as a safety layer that makes sure the `RawRwLock` is used correctly. But I realize the current impl is not correct either. We need to count the readers. Currently if there are 5 readers and one of them calls `unlock` the entire lock state is set to `UNLOCKED` in our wrapper, but the internal `RawRwLock` will of course still be read-locked with 4 readers.

I saw your other PR. Would be great if that was merged, then the majority of these functions would not be a concern in this PR at all.

Reply:
It doesn't make sense to me to replicate the internal `RawRwLock` state. At that point you're basically recreating an R/W lock. Why not just add a `pub(crate) fn unlock` to `RawRwLock`?

Reply:
It does not make much sense to me to complicate things and change the public API of a well functioning crate with a nice API to cover this corner case that can very well be covered by adding 8 extra bytes of state and a lot less code than it would take to implement `RawRwLock::unlock`. Does it matter that the libunwind lock becomes 16 instead of 8 bytes, and that it has a few extra atomic operations? Will libunwind use the lock so heavily that the performance and memory usage is that critical?

Publicly exposing `unlock_shared`, `unlock_exclusive` and `unlock` from `RawRwLock` will make the API of the type harder to understand and very likely more people would use it in the wrong way.

It can also be improved later. If I solve the problem at hand with an extra type now this PR can progress.

Reply:
> Does it matter that the libunwind lock becomes 16 instead of 8 bytes, and that it has a few extra atomic operations? Will libunwind use the lock so heavily that the performance and memory usage is that critical?

No to both.

Note that I was not suggesting changing the public API.

Reply:
I pushed an updated version to the same testing branch. Use the same link above if you want to check the code.

Reply:
Oh now I see what you did. I completely misread the patch and thought you were putting these in the SGX module of parking_lot, but you're putting them in std. In that case, maybe it's best to just collect all the functions into a separate `sgx/unwind.rs`, now that we no longer need to override rwlock privacy.

Reply:
I moved them to a separate `unwind.rs` on the test branch. Are you referring to the "`Odrering`" spelling error being the test failure? I have fixed that now. Running `dist-various-2` as we speak to verify if the testing branch is currently compiling or not.

Reply:
No, at least you're mixing the `RawRwLock` and `RwLock` types. The dist-various-2 builder doesn't build the tests unfortunately...

Reply:
Oops. Pushed an update. Could you help me test if the tests compile now?

Reply:
@jethrogb Should you maybe close #59624 and I'll pull in your changes to `src/librustc_target/spec/x86_64_fortanix_unknown_sgx.rs` in this PR? Or what's your thought?

Reply:
I'd like to land those changes ASAP, I don't know how close this PR is to being accepted. I don't mind putting them all back into a single place, I didn't do that in my PR because there was an obvious place to put each of them.

Reply:
I pushed the things I showed you from the testing branch into this PR now. Please tell me if the tests still don't build for SGX. Or inform me how to build/run them properly :)

Reply:
Oh yeah sorry I was working on testing this but my CI is down and I got distracted. You can run something like 
```
touch libunwind.a
X86_FORTANIX_SGX_LIBS=$PWD ./x.py test --stage 1 --target x86_64-fortanix-unknown-sgx src/libstd --no-doc
```
This will fail to link but it should at least get you past the compiler errors stage.

```
error[E0433]: failed to resolve: use of undeclared type or module `Atomicbool`
   --> src/libstd/sys/sgx/unwind.rs:123:34
    |
123 |                 is_write_locked: Atomicbool::new(false),
    |                                  ^^^^^^^^^^ use of undeclared type or module `Atomicbool`

error: unused imports: `Layout`, `Ordering`, `lock_api::RawRwLock as _`, `self`, `slice`, `str`
  --> src/libstd/sys/sgx/unwind.rs:10:13
   |
10 |     alloc::{self, Layout},
   |             ^^^^  ^^^^^^
11 |     lock_api::RawRwLock as _,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
12 |     parking_lot::RawRwLock,
13 |     slice, str,
   |     ^^^^^  ^^^
14 |     sync::atomic::{AtomicBool, Ordering},
   |                                ^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`
```

Reply:
Oh, lol. Coding without compiling is 100% doomed to fail on the smallest of things 🤦‍♂️ Now I have pushed something that actually builds to the testing branch. Will push it here when I have something more to add as well. Don't want to stress the CI more than needed. Thanks for the help.